### PR TITLE
Support date_trunc

### PR DESCRIPTION
**Pull request by pdobacz**
A lot of rough edges on this one, so opening as draft.

I'm not happy with entire implementation of `primary_arg`, but the best alternative I see is to add a rich `function_infos` dictionary in `oid_cache`, populate it during startup/first use and look up things like `is_cast` and `primary_arg` by OID there. However, previously we rejected the "populate on startup" part so I didn't venture there yet.

Second point: I abstracted away the part to get seed material by type, but now this involves a lot of `pallocs`. I checked with the `stress.sql` test and query to `taxi` and the timing is the same.

Lastly, I didn't go down the normalization path, as I'm not sure it's worth the extra code and abstractions at this point.

### CODE REVIEW

File: src/query/allowed_objects.c [hunk 1]
```diff
@@ -14,18 +14,49 @@ static const char *const g_allowed_casts[] = {
     "ftod", "dtof",
     "int4_numeric", "float4_numeric", "float8_numeric",
     "numeric_float4", "numeric_float8",
+    "date_timestamptz",
     /**/
 };
 
-static const char *const g_allowed_builtins[] = {
+typedef struct FunctionWithName
```
Comment (cristianberneanu) [inline src/query/allowed_objects.c:21]: Maybe drop the `With`?

**Review by cristianberneanu — APPROVED** [review]

File: src/query/allowed_objects.c (hunk 1 above)
Comment (pdobacz) [reply src/query/allowed_objects.c:21]: hm, that would make these types `FunctionName` and `FunctionOid`. Could work, but I've had a late thought: `FunctionByName` and `FunctionByOid` - what we actually are achieving here is a mapping, just with linear lookup. Would these work?

And a more general question, to make sure - I'm assuming the :heavy_check_mark: means you prefer the current version with the arrays to the alternative with a runtime-populated dictionary?

Comment (cristianberneanu) [reply src/query/allowed_objects.c:21]: `FunctionByName` and `FunctionByOid` sound good to me.

For the small collections we have, I think linear searches are fine. If we ever find they are a bottleneck, we can pre-sort the arrays and do a binary search instead, to get a speed boost.

File: src/query/allowed_objects.c [hunk 2]
```diff
@@ -14,18 +14,49 @@ static const char *const g_allowed_casts[] = {
...
+{
+  const char *name;
+  int primary_arg;
+} FunctionWithName;
+
+typedef struct FunctionWithOid
+{
+  Oid funcid;
+  int primary_arg;
+} FunctionWithOid;
```
Comment (edongashi) [inline src/query/allowed_objects.c:31]: Can we merge these 2 structs and initialize missing fields during oid cache build?

I.e. we can have `("some_func", InvalidOid, 0)` OR `(NULL, F_NUMERIC_ROUND_INT, 0)`. 
During init, we fill out the 0-OID from the Name or we fill out the NULL-Name from OID (the latter might not be necessary?).

Would this simplify the lookup logic? Also would get rid of the func name lookups we do during validation.

File: src/query/allowed_objects.c [hunk 3]
```diff
@@ -14,18 +14,49 @@ static const char *const g_allowed_casts[] = {
...
+
+static const FunctionWithName g_allowed_builtins[] = {
     /* rounding casts */
-    "ftoi2", "ftoi4", "ftoi8", "dtoi2", "dtoi4", "dtoi8", "numeric_int4",
+    (FunctionWithName){.name = "ftoi2", .primary_arg = 0},
```
Comment (edongashi) [inline src/query/allowed_objects.c:35]: Does dropping `.primary_arg = 0` cause it be zero anyway?

File: src/query/allowed_objects.c (hunk 1 above)
Comment (pdobacz) [reply src/query/allowed_objects.c:21]: to clarify - my focus here is more on code clarity, not the speed of lookups. But if that's fine by you as well, let's keep the current approach.

File: src/query/allowed_objects.c (hunk 2 above)
Comment (pdobacz) [reply src/query/allowed_objects.c:31]: The part "initialize during oid cache build" is the runtime population we were originally avoiding here. If we change the preferred approach to allow runtime population, this sounds like a viable alternative to the `Oid -> FunctionInfo` mapping I had in mind

File: src/query/anonymization.c [hunk 4]
```diff
@@ -534,6 +534,39 @@ static void append_seed_material(
   strcpy(existing_material + existing_material_length, new_material);
 }
 
+static char *datum_seed_material(Oid type, Datum value, bool is_null)
+{
+  if (is_null)
+  {
+    char *value_as_string = palloc(sizeof(char) * (4 + 1));
+    strcpy(value_as_string, "NULL");
```
Comment (edongashi) [inline src/query/anonymization.c:542]: Will this also copy the final `\0`?

File: src/query/allowed_objects.c (hunk 1 above)
Comment (cristianberneanu) [reply src/query/allowed_objects.c:21]: This being C, I think the array version is the clearer one :)

File: src/query/allowed_objects.c (hunk 3 above)
Comment (pdobacz) [reply src/query/allowed_objects.c:35]: it does indeed, but it's not a meaningful default in here, so I'm reluctant to leverage this

Comment (edongashi) [reply src/query/allowed_objects.c:35]: Yes, `primary_arg` should always be explicit. The zero trick will work well for `.name` or `.oid` if they are mutually exclusive.

**Review by cristianberneanu — APPROVED** [review]

**Review by cristianberneanu — APPROVED** [review]

**Pull request merged**